internal/argocd/cluster/cluster.go

```diff
@@ -39,67 +44,154 @@ func (m *Manager) UpdateClusterConnectionInfo(agentName, status string, t time.T
+			}
+			continue
+		}
 
-		if err := common.SetOptionalRedisPasswordFromKubeConfig(m.ctx, m.kubeclient, m.namespace, redisOptions); err != nil {
-			log().Errorf("Failed to fetch & set redis password for namespace %s: %v", m.namespace, err)
+		// Re-save same info.
+		if err := setClusterInfo(m.ctx, m.kubeclient, m.namespace, cluster.Server, m.redisAddress,
+			agentName, cluster.Name, m.redisCompressionType, clusterInfo); err != nil {
+			log().Errorf("failed to refresh connection info in cluster: '%s' mapped with agent: '%s'. Error: %v", cluster.Name, agentName, err)
+			continue
 		}
-		redisClient := redis.NewClient(redisOptions)
+	}
+}
 
-		cache := appstatecache.NewCache(cacheutil.NewCache(
-			cacheutil.NewRedisCache(redisClient, time.Minute, m.redisCompressionType)), time.Minute)
+// SetClusterCacheStats updates cluster cache info with Application, Resource and API counts in principal.
+// This is called when principal receives clusterCacheInfoUpdate event from agent.
+func (m *Manager) SetClusterCacheStats(clusterInfo appv1.ClusterInfo, agentName string) {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
 
-		// fetch latest info
-		clusterInfo := &appv1.ClusterInfo{}
-		if err := cache.GetClusterInfo(cluster.Server, clusterInfo); err != nil {
-			log().Errorf("Failed to get connection info from Cluster: '%s' mapped with Agent: '%s'. Error: %v", cluster.Name, agentName, err)
-			return
-		}
+	// Check if we have a mapping for the requested agent.
+	cluster := m.mapping(agentName)
+	if cluster == nil {
+		log().Errorf("agent %s is not mapped to any cluster", agentName)
+		return
+	}
+
+	// Get existing cluster info to preserve cluster connection status
+	existingClusterInfo, err := GetClusterInfo(m.ctx, m.kubeclient, m.namespace, cluster.Server, m.redisAddress, m.redisCompressionType)
+	if err != nil && !errors.Is(err, cacheutil.ErrCacheMiss) {
+		log().Errorf("failed to get existing cluster info for cluster: '%s' mapped with agent: '%s'. Error: %v", cluster.Name, agentName, err)
+		// Continue with default values if we can't get existing info
+	}
 
-		// re-save same info
-		if err := cache.SetClusterInfo(cluster.Server, clusterInfo); err != nil {
-			log().Errorf("Failed to refresh connection info in Cluster: '%s' mapped with Agent: '%s'. Error: %v", cluster.Name, agentName, err)
-			return
+	// Create new cluster info with cache stats
+	newClusterInfo := &appv1.ClusterInfo{
+		ApplicationsCount: clusterInfo.ApplicationsCount,
+		CacheInfo: appv1.ClusterCacheInfo{
+			APIsCount:      clusterInfo.CacheInfo.APIsCount,
+			ResourcesCount: clusterInfo.CacheInfo.ResourcesCount,
+		},
+	}
+
+	// Preserve existing cluster connection status
+	if existingClusterInfo != nil {
+		newClusterInfo.ConnectionState = existingClusterInfo.ConnectionState
+		if existingClusterInfo.CacheInfo.LastCacheSyncTime != nil {
+			newClusterInfo.CacheInfo.LastCacheSyncTime = existingClusterInfo.CacheInfo.LastCacheSyncTime
```

Comment:
Can you explain the logic of these lines? From my quick read, it seems like the newer (less stale?) value would be overwritten by the older value (in all cases).